src/equations/shallow_water_2d.jl

```diff
@@ -0,0 +1,564 @@
+
+  c = equations.gravity * sqrt(h)
+  return abs(v1) + c, abs(v2) + c
+end
+
+
+# Helper function to extract the velocity vector from the conservative variables
+@inline function velocity(u, equations::ShallowWaterEquations2D)
+  _, v1, v2, _ = cons2prim(u, equations)
+  return SVector(v1, v2)
+end
+
+
+# Convert conservative variables to primitive
+@inline function cons2prim(u, equations::ShallowWaterEquations2D)
+  h, h_v1, h_v2, b = u
+
+  H = h + b
+  v1 = h_v1 / h
+  v2 = h_v2 / h
+  return SVector(H, v1, v2, b)
+end
+
+
+# Convert conservative variables to entropy
+# Note, only the first three are the entropy variables, the fourth entry still
+# just carries the bottom topography values for convenience
+@inline function cons2entropy(u, equations::ShallowWaterEquations2D)
+  h, h_v1, h_v2, b = u
+
+  v1 = h_v1 / h
+  v2 = h_v2 / h
+  v_square = v1^2 + v2^2
+
+  w1 = equations.gravity * (h + b) - 0.5 * v_square
+  w2 = v1
+  w3 = v2
+  return SVector(w1, w2, w3, b)
+end
+
+
+# Convert entropy variables to conservative
+@inline function entropy2cons(w, equations::ShallowWaterEquations2D)
+  w1, w2, w3, b = w
+
+  h = (w1 + 0.5 * (w2^2 + w3^2)) / equations.gravity - b
+  h_v1 = h * w2
+  h_v2 = h * w3
+  return SVector(h, h_v1, h_v2, b)
+end
+
+
+# Convert primitive to conservative variables
+@inline function prim2cons(prim, equations::ShallowWaterEquations2D)
+  H, v1, v2, b = prim
+
+  h = H - b
+  h_v1 = h * v1
+  h_v2 = h * v2
+  return SVector(h, h_v1, h_v2, b)
+end
+
+
+@inline function waterheight(u, equations::ShallowWaterEquations2D)
+  return u[1]
+end
+
+
+@inline function pressure(u, equations::ShallowWaterEquations2D)
+  h = u[1]
+  p = 0.5 * equations.gravity * h^2
+  return p
+end
+
+
+@inline function waterheight_pressure(u, equations::ShallowWaterEquations2D)
+  h = u[1]
+  h_times_p = 0.5 * equations.gravity * h^3
+  return h_times_p
+end
+
+
+# Entropy function for the shallow water equations is the total energy
+@inline entropy(cons, equations::ShallowWaterEquations2D) = energy_total(cons, equations)
+
+
+# Calculate total energy for a conservative state `cons`
+@inline function energy_total(cons, equations::ShallowWaterEquations2D)
+  h, h_v1, h_v2, b = cons
+
+  e = (h_v1^2 + h_v2^2) / (2 * h) + 0.5 * equations.gravity * h^2 + equations.gravity * h * b
+  return e
+end
+
+
+# Calculate kinetic energy for a conservative state `cons`
+@inline function energy_kinetic(u, equations::ShallowWaterEquations2D)
+  h, h_v1, h_v2, _ = u
+  return (h_v1^2 + h_v2^2) / (2 * h)
+end
+
+
+# Calculate potential energy for a conservative state `cons`
+@inline function energy_internal(cons, equations::ShallowWaterEquations2D)
+  return energy_total(cons, equations) - energy_kinetic(cons, equations)
+end
+
+
+# Calculate the error for the "lake-at-rest" test case where H = h+b should
+# be a constant value over time
+@inline function lake_at_rest_error(u, equations::ShallowWaterEquations2D)
+  return abs(equations.H0 - (u[1] + u[4]))
```

Comment:
Maybe instead of arbitrary indexing, use something more canonical?
```suggestion
  return abs(equations.H0 - (waterheight(u, equations)] + bottomheight(u, equations)))
```
or similar?

Reply:
hmm, maybe. This would require making another function called something like `bottomheight` that would extract `u[4]` in a similar fashion to how `waterheight` extracts `u[1]`. Maybe I could just add a comment about why I use this indexing?

Reply:
I am also Ok with leaving it as it is; I just know that @ranocha is not a big fan of explicit indexing since it is hard to remember for a non-regular user what the indices mean.